taropsbt/interface.go

```diff
@@ -0,0 +1,170 @@
+// be merged into a single BTC transaction. Thus, each asset state transfer is
+// represented in a virtual TX and multiple asset state transfers can be
+// anchored within a single BTC transaction.
+//
+// TODO(guggero): Actually support merging multiple virtual transactions into a
+// single BTC transaction.
+type VPacket struct {
+	// Input is the asset input that is being spent.
+	//
+	// TODO(guggero): Support spending multiple inputs.
+	Input *VInput
+
+	// Outputs is the list of new asset outputs that are created by the
+	// virtual transaction. By convention the output at index 0 is the
+	// change output for asset change or the asset tombstone in case of a
+	// non-interactive full value send.
+	Outputs []*VOutput
+
+	// ChainParams are the Taro chain parameters that are used to encode and
+	// decode certain contents of the virtual packet.
+	ChainParams *address.ChainParams
+}
+
+// SetInputAsset sets the input asset that is being spent.
+func (p *VPacket) SetInputAsset(a *asset.Asset) {
+	p.Input.asset = a
+}
+
+// Anchor is a struct that contains all the information about an anchor output.
+type Anchor struct {
+	// Value is output value of the anchor output.
+	Value btcutil.Amount
+
+	// PkScript is the output script of the anchor output.
+	PkScript []byte
+
+	// SigHashType is the signature hash type that should be used to sign
+	// the anchor output spend.
+	SigHashType txscript.SigHashType
+
+	// InternalKey is the internal key of the anchor output that the input
+	// is spending the asset from.
+	InternalKey *btcec.PublicKey
+
+	// MerkleRoot is the root of the tap script merkle tree that also
+	// contains the Taro commitment of the anchor output.
+	MerkleRoot []byte
+
+	// Bip32Derivation is the BIP32 derivation of the anchor output's
+	// internal key.
+	//
+	// TODO(guggero): Do we also want to allow multiple derivations to be
+	// specified here? That would allow us to specify multiple keys involved
+	// in MuSig2 for example. Same for the Taproot derivation below.
+	Bip32Derivation *psbt.Bip32Derivation
+
+	// TrBip32Derivation is the Taproot BIP32 derivation of the anchor
+	// output's internal key.
+	TrBip32Derivation *psbt.TaprootBip32Derivation
+}
+
+// VInput represents an input to a virtual asset state transition transaction.
+type VInput struct {
+	// PInput is the embedded default PSBT input struct that is used for
+	// asset related input data.
+	psbt.PInput
+
+	// PrevID is the asset previous ID of the asset being spent.
+	PrevID asset.PrevID
+
+	// Anchor contains the information about the BTC level anchor
+	// transaction that committed to the asset being spent.
+	Anchor Anchor
+
+	// TapscriptSibling is the tapscript sibling of this asset. This will
+	// usually be blank.
+	TapscriptSibling []byte
+
+	// asset is the full instance of the asset being spent. It is not
+	// exported because the assets script key must be encoded in the PSBT
+	// input struct for the signing to work correctly.
+	asset *asset.Asset
```

Comment:
May have been lost in some of the comments in the original mega PR but: don't we also need an SMT reveal here along with the tapscript reveal? This then permits verification that the asset actually lives within the anchor above. Or is the assumption that it'll be fetched out of band with a Universe API?

Reply:
w/ the above, we can also potentially break out portions of the `proof.Proof` struct to reside here as well.

Reply:
Good point, I did miss that part in the previous PR. Since the proofs shouldn't be too large (as compared to the full input commitment), I'll try adding it. And if that will simplify the proof creation, even better!

Reply:
Yeah this is just the inclusion proof that the input asset actually exists (not full proof history). With that, I don't think we need to worry about any length limits for the PSBT, assuming file/direct/encoded transfer.